ogr/ogrsf_frmts/miramon/mm_gdal_functions.c

```diff
@@ -0,0 +1,2801 @@
+                                     const char *pszRelFile)
+{
+    MM_BYTE variable_byte;
+    FILE_TYPE *pf;
+    unsigned short int ushort;
+    MM_EXT_DBF_N_FIELDS nIField;
+    MM_FIRST_RECORD_OFFSET_TYPE offset_primera_fitxa;
+    MM_FIRST_RECORD_OFFSET_TYPE offset_fals = 0;
+    MM_BOOLEAN incoherent_record_size = FALSE;
+    MM_BYTE un_byte;
+    MM_BYTES_PER_FIELD_TYPE_DBF bytes_per_camp;
+    MM_BYTE tretze_bytes[13];
+    MM_FIRST_RECORD_OFFSET_TYPE offset_possible;
+    MM_BYTE n_queixes_estructura_incorrecta = 0;
+    MM_FILE_OFFSET offset_reintent = 0;  // For retrying
+    char cpg_file[MM_CPL_PATH_BUF_SIZE];
+    char *pszDesc;
+    char section[MM_MAX_LON_FIELD_NAME_DBF + 25];  // TAULA_PRINCIPAL:field_name
+    GUInt32 nRecords;
+    char *pszString;
+
+    if (!szFileName)
+        return 1;
+
+    strcpy(pMMBDXP->szFileName, szFileName);
+    strcpy(pMMBDXP->ReadingMode, "rb");
+
+    if ((pMMBDXP->pfDataBase = fopen_function(pMMBDXP->szFileName,
+                                              pMMBDXP->ReadingMode)) == nullptr)
+        return 1;
+
+    pf = pMMBDXP->pfDataBase;
+
+    fseek_function(pf, 0, SEEK_SET);
+    /* ====== Header reading (32 bytes) =================== */
+    offset_primera_fitxa = 0;
+
+    if (1 != fread_function(&(pMMBDXP->dbf_version), 1, 1, pf) ||
+        1 != fread_function(&variable_byte, 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->month), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->day), 1, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    if (1 != fread_function(&nRecords, 4, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    if (1 != fread_function(&offset_primera_fitxa, 2, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    pMMBDXP->year = (short)(1900 + variable_byte);
+reintenta_lectura_per_si_error_CreaCampBD_XP:
+
+    if (n_queixes_estructura_incorrecta > 0)
+    {
+        if (!MM_ES_DBF_ESTESA(pMMBDXP->dbf_version))
+        {
+            offset_fals = offset_primera_fitxa;
+            if ((offset_primera_fitxa - 1) % 32)
+            {
+                for (offset_fals = (offset_primera_fitxa - 1);
+                     !((offset_fals - 1) % 32); offset_fals--)
+                    ;
+            }
+        }
+    }
+    else
+        offset_reintent = ftell_function(pf);
+
+    if (1 != fread_function(&ushort, 2, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->reserved_1), 2, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->transaction_flag), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->encryption_flag), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->dbf_on_a_LAN), 12, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    if (MM_ES_DBF_ESTESA(pMMBDXP->dbf_version))
+    {
+        memcpy(&pMMBDXP->nRecords, &nRecords, 4);
+        memcpy(((char *)&pMMBDXP->nRecords) + 4, &pMMBDXP->dbf_on_a_LAN, 4);
```

Comment:
would be cleaner to copy in a nRecordsLow32Bits and nRecordsHigh64Bits, and combine them later with ``nRecords = ((GUInt64)nRecordsHigh64Bits << 32) | nRecordsLow32Bits``

Reply:
This must be clearer now: 
if (MM_ES_DBF_ESTESA(pMMBDXP->dbf_version))
    {
        GUInt32 nRecordsHigh32Bits;

        // Getting 4 bytes of the 8 bytes variable
        memcpy(&nRecordsHigh32Bits, &pMMBDXP->dbf_on_a_LAN, 4);

        // Getting other 4 bytes of the 8 bytes variable
        // The cast to GUInt64 of the high 32 bits is important to
        // make sure the left bit shift is done correctly
        pMMBDXP->nRecords = ((GUInt64)nRecordsHigh32Bits << 32) | nRecordsLow32Bits;
    }
    else
        pMMBDXP->nRecords = nRecordsLow32Bits;

Reply:
yes looks good